compiler/optimizer/abstractinterpreter/InliningMethodSummary.cpp

```diff
@@ -0,0 +1,177 @@
+
+   if (_optsByArg.size() <= argPos || _optsByArg[argPos] == NULL || _optsByArg[argPos]->size() == 0)
+      return 0;
+
+   uint32_t benefit = 0;
+
+   for (size_t i = 0; i < _optsByArg[argPos]->size(); i ++)
+      {
+      TR::PotentialOptimizationPredicate* predicate = (*_optsByArg[argPos])[i];
+      if (predicate->test(arg))
+         {
+         benefit += 1;
+         }
+      }
+
+   return benefit;
+   }
+
+void TR::InliningMethodSummary::trace(TR::Compilation* comp)
+   {
+   traceMsg(comp, "Inlining Method Summary:\n");
+
+   if (_optsByArg.size() == 0)
+      {
+      traceMsg(comp, "EMPTY\n\n");
+      return;
+      }
+
+   for (size_t i = 0; i < _optsByArg.size(); i ++)
+      {
+      if (_optsByArg[i] != NULL)
+         {
+         for (size_t j = 0; j < _optsByArg[i]->size(); j ++)
+            {
+            TR::PotentialOptimizationPredicate* predicate = (*_optsByArg[i])[j];
+
+            traceMsg(comp, "%s @%d for Argument %d ", predicate->getName(), predicate->getBytecodeIndex(), i);
+            predicate->trace(comp);
+            traceMsg(comp, "\n");
+            }
+         }
+      }
+
+   }
+
+
+void TR::InliningMethodSummary::addPotentialOptimizationByArgument(TR::PotentialOptimizationPredicate* predicate, uint32_t argPos)
+   {
+   if (_optsByArg.size() <= argPos)
+      _optsByArg.resize(argPos + 1);
+
+   if (!_optsByArg[argPos])
+       _optsByArg[argPos] = new (region()) PredicateContainer(region());
+
+   _optsByArg[argPos]->push_back(predicate);
+   }
+
+const char* TR::PotentialOptimizationPredicate::getName()
+   {
+   switch (_kind)
+      {
+      case Kind::BranchFolding:
+         return "Branch Folding";
+      case Kind::CheckCastFolding:
+         return "CheckCast Folding";
+      case Kind::NullCheckFolding:
+         return "NullCheck Folding";
+      case Kind::InstanceOfFolding:
+         return "InstanceOf Folding";
+      default:
+         TR_ASSERT_FATAL(false, "Unexpected Kind");
+      }
+   }
+
+bool TR::PotentialOptimizationVPPredicate::holdPartialOrderRelation(TR::VPConstraint* valueConstraint, TR::VPConstraint* testConstraint)
+   {
+   if (testConstraint->asIntConstraint()) // partial relation for int constraint
+      {
+      if (testConstraint->getLowInt() <= valueConstraint->getLowInt() && testConstraint->getHighInt() >= valueConstraint->getHighInt())
+         return true;
+      else
+         return false;
+      }
+   else if (testConstraint->asClassPresence()) // partial relation for nullness
+      {
+      if (testConstraint->isNonNullObject() && valueConstraint->isNonNullObject())
+         return true;
+      else if (testConstraint->isNullObject() && valueConstraint->isNullObject())
+         return true;
+      else
+         return false;
+      }
+   else if (testConstraint->asClassType()) // testing for checkcast
+      {
+      TR_ASSERT_FATAL(valueConstraint->getClassType()->asResolvedClass(), "valueConstraint unexpectedly admits unresolved class type");
```

Comment:
Sorry, it was my mistake for the assertion. Our intention for the assertion was to check the type we cast to is a valid type, so the object reference pushed to the top of the stack must be of a type specified in the constant pool. And you are right, we should make assertion with the `testConstraint`, rather than the `valueConstraint`. I have changed it now.
I used the code that you suggested below to return `false` if `testConstraint->isClassObject() == TR_yes`, since we do not want to handle reflective object.